src/proof_utils.rs

```diff
@@ -0,0 +1,100 @@
+// Copyright 2023 The Fuchsia Authors
+//
+// Licensed under a BSD-style license <LICENSE-BSD>, Apache License, Version 2.0
+// <LICENSE-APACHE or https://www.apache.org/licenses/LICENSE-2.0>, or the MIT
+// license <LICENSE-MIT or https://opensource.org/licenses/MIT>, at your option.
+// This file may not be copied, modified, or distributed except according to
+// those terms.
+
+//! This module exists to hold this doc comment, which provides lemmas which can
+//! be used in soundness proofs.
+//!
+//! # Definitions
+//!
+//! ## transparent-layout-validity
+//!
+//! A type, `T`, has the property `transparent-layout-validity(U)` if the
+//! following all hold:
+//! - `T` and `U` have the same alignment
+//! - For all `t: *const T`, `let u = t as *const U` is valid and
+//!   `size_of_val_raw(t) == size_of_val_raw(u)`.
+//! - For all `u: *const U`, `let t = *const T` is valid and `size_of_val_raw(u)
+//!   == size_of_val_raw(t)`.
+//! - For all `(t, u): (*const T, *const U)` where `size_of_val_raw(t) ==
+//!   size_of_val_raw(u)`:
+//!   - `t` and `u` refer to `UnsafeCell`s at the same byte ranges.
+//!   - If `*t` contains a valid `T`, that implies that `*u` contains a valid
+//!     `U`.
+//!   - If `*u` contains a valid `U`, that implies that `*t` contains a valid
+//!     `T`.
+//!
+//! # Axioms
+//!
+//! These are statements which are not technically logically bulletproof, but
+//! capture the way that language is used in practice in the Rust Reference and
+//! standard library documentation.
+//!
+//! ## axiom-transparent-layout-validity
+//!
+//! Given types `T` and `U`, the phrase "`T` is guaranteed to have the same
+//! layout and bit validity as `U`" is taken to imply that `T` has the property
+//! `transparent-layout-validity(U)`.
+//!
+//! The term "layout" is used in Rust documentation to refer to a type's size
+//! and alignment and the sizes, alignments, and byte offsets of each of the
+//! type's fields. In practice, phrases like the above are only ever used in
+//! contexts where the following additional properties also hold:
+//! - `T` and `U` have the same vtable kinds. `T`'s and `U`'s pointer metadata
+//!   is such that raw pointer casts preserve size and field placement.
+//! - `T` and `U` have `UnsafeCell`s at the same byte ranges.
+//!
+//! ## axiom-repr-transparent-layout-validity
+//!
+//! Given types `T` and `U`, if `T` is a `#[repr(transparent)]` struct with a
+//! `pub` field of type `U`, and `T` does not contain any other fields, then
```

Comment:
Is "`T` does not contain any other fields" a necessary precondition? A `repr(transparent)` type may contain at most one field with non-trivial size and alignment. The addition of fields with trivial size (`0`) and alignment (`1`) cannot  impact the conditions of `TRANSPARENT_LAYOUT_VALIDIDTY`.

Reply:
My only concern is that formally defining /which/ field is non-zero-sized would be annoying, and I'm not sure how to articulate it. Obviously you can just say "the only non-zero-sized field", but that kicks the can down the road: how do you know, at the proof site, which fields are zero sized? Right now, we only use this for `Wrapped`, which satisfies this more restricted precondition.